### ISSUE TITLE
[Proposal] Expose onChanged property of Searchfield

### ISSUE THREAD

**Issue by TomJFox** [maheshmnj/searchfield#78]
**Is your feature request related to a problem? Please describe.**
I would love to have a way to pass as a property a minimum number of characters which need to be entered before the overlay is shown with search results.

**Describe the solution you'd like**
Add a property (eg. minimumSuggestionLength) and add a check at the beginning of onChanged in the TextFormField which simply returns if the query.length is less than or equal to the property.

**Describe alternatives you've considered**
This was the only way I could think of that this could be accomplished.

**Additional context**

```
        onChanged: (query) {
          final searchResult = <SearchFieldListItem<T>>[];

          if (query.length <= widget.minimumSuggestionLength) {
            return;
          }

```

**Comment by maheshmnj** [comment]
Hello @TomJFox, Thanks for filing the issue. You should be able to achieve that with a couple of lines of code by adding a text property manually below the searchfield, Would this work? 

declare max length
```
int minimumSuggestionLength = 5;
```
// add a listener to be able to update state on text changed
```dart
searchController!.addListener(() { 
     setState((){});
    });
```  

```dart
onChanged: (query) {
  searchResult = <SearchFieldListItem<T>>[];

  if (query.length <= minimumSuggestionLength) {
      return;
}
```

**Comment by TomJFox** [comment]
@maheshmnj I am having trouble getting this to work.  Would it be possible to share how you would update this simple example?

```
class SearchFieldTest extends StatefulWidget {
  const SearchFieldTest({
    Key? key,
    this.width,
    this.height,
  }) : super(key: key);

  final double? width;
  final double? height;

  @override
  _SearchFieldTestState createState() => _SearchFieldTestState();
}

class _SearchFieldTestState extends State<SearchFieldTest> {
  @override
  Widget build(BuildContext context) {
    return SearchField(
      key: const Key('searchfield'),
      suggestionItemDecoration: BoxDecoration(
          borderRadius: BorderRadius.circular(8),
          shape: BoxShape.rectangle,
          border: Border.all(
              color: Colors.transparent, style: BorderStyle.solid, width: 1.0)),
      suggestionState: Suggestion.hidden,
      suggestions: ['ABC', 'DEF', 'def']
          .map((e) => SearchFieldListItem<String>(e))
          .toList(),
    );
  }
}

```

**Comment by maheshmnj** [comment]
Ah sorry for the confusion I thought Searchfield had `onChanged` property, perhaps we should expose onChanged property of the TextField to SearchField.

then this code would work
```dart
class SearchFieldTest extends StatefulWidget {
  const SearchFieldTest({
    Key? key,
    this.width,
    this.height,
  }) : super(key: key);

  final double? width;
  final double? height;

  @override
  _SearchFieldTestState createState() => _SearchFieldTestState();
}

class _SearchFieldTestState extends State<SearchFieldTest> {

  @override
  void dispose() {
    searchController.dispose();
      super.dispose();
    }

    @override
    void initState() {
        // TODO: implement initState
        super.initState();
        searchController.addListener((){
          setState(() {
                      
                    });
        });
      }

  final searchController = TextEditingController();
  int minimumSuggestionLength = 10;
  @override
  Widget build(BuildContext context) {
    return SearchField(
      key: const Key('searchfield'),
      controller: searchController,
      onChanged: (query){
        if (query.length <= minimumSuggestionLength) {
            return;
      },
      suggestionItemDecoration: BoxDecoration(
          borderRadius: BorderRadius.circular(8),
          shape: BoxShape.rectangle,
          border: Border.all(
              color: Colors.transparent, style: BorderStyle.solid, width: 1.0)),
      suggestionState: Suggestion.hidden,
      suggestions: ['ABC', 'DEF', 'def']
          .map((e) => SearchFieldListItem<String>(e))
          .toList(),
    );
  }
}
```

Alternatively I tried something like this and its performance is severely bad if you do setState since it rebuilds the overlay but with ValueListenableBuilder it should be fine.

```dart
import 'package:flutter/material.dart';
import 'package:searchfield/searchfield.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter App',
      theme: ThemeData(
        colorSchemeSeed: Colors.indigo,
        useMaterial3: true,
        brightness: Brightness.light,
      ),
      darkTheme: ThemeData(
        colorSchemeSeed: Colors.blue,
        useMaterial3: true,
        brightness: Brightness.dark,
      ),
      home: SearchFieldSample(),
      debugShowCheckedModeBanner: false,
    );
  }
}

class SearchFieldSample extends StatefulWidget {
  const SearchFieldSample({Key? key}) : super(key: key);

  @override
  State<SearchFieldSample> createState() => _SearchFieldSampleState();
}

class _SearchFieldSampleState extends State<SearchFieldSample> {
  List<String> suggestions =
      [
    'United States',
    'Germany',
    'Washington',
    'Paris',
    'Jakarta',
    'Australia',
    'India',
    'Czech Republic',
    'Lorem Ipsum',
  ];
  int suggestionsCount = 2;
  final focus = FocusNode();
  @override
  Widget build(BuildContext context) {
    return Scaffold(
        appBar: AppBar(
          title: Text('Dynamic sample Demo'),
        ),
        floatingActionButton: FloatingActionButton(
          onPressed: () {
            setState(() {
              suggestionsCount++;
              suggestions.add('suggestion $suggestionsCount');
            });
          },
          child: Icon(Icons.add),
        ),
        body: Center(
          child: SearchFieldTest()
        ));
  }
}

class SearchFieldTest extends StatefulWidget {
  const SearchFieldTest({
    Key? key,
    this.width,
    this.height,
  }) : super(key: key);

  final double? width;
  final double? height;

  @override
  _SearchFieldTestState createState() => _SearchFieldTestState();
}

class _SearchFieldTestState extends State<SearchFieldTest> {
  @override
  void dispose() {
    searchController.dispose();
    errorNotifier.dispose();
    super.dispose();
  }

  @override
  void initState() {
    super.initState();
      searchController.addListener(() {
        oldText = searchController.text;
        if (oldText.length < minimumSuggestionLength) {
          errorNotifier.value = false;
          return;
        }else{
          errorNotifier.value = true;
        }
      });
  }
  final ValueNotifier<bool> errorNotifier = ValueNotifier<bool>(false);
  String oldText = "";
  final searchController = TextEditingController();
  int minimumSuggestionLength = 5;
  @override
  Widget build(BuildContext context) {
    return Column(
      children: [
        SearchField(
          key: const Key('searchfield'),
          controller: searchController,
          // suggestionDirection: SuggestionDirection.up,
          suggestionItemDecoration: BoxDecoration(
              borderRadius: BorderRadius.circular(8),
              shape: BoxShape.rectangle,
              border: Border.all(
                  color: Colors.transparent,
                  style: BorderStyle.solid,
                  width: 1.0)),
          suggestionState: Suggestion.hidden,
          suggestions: ['ABC', 'DEF', 'def']
              .map((e) => SearchFieldListItem<String>(e))
              .toList(),
        ),
        ValueListenableBuilder<bool>(
          valueListenable: errorNotifier,
        builder: (x,showError,child){
          return (showError)? Text("This is a custom text")
          : SizedBox.shrink();
        },
        )
      ],
    );
  }
}

```

**Comment by maheshmnj** [comment]
I am guessing it should be as simple as adding an `onChanged` property and calling it here

https://github.com/maheshmnj/searchfield/blob/1f5f20de648abe9638a3261f674104f9c28f1e18/lib/src/searchfield.dart#L613-L618

Feel free to make a PR if you feel interested and I would be happy to review.

**Comment by arifikhsan** [comment]
yes. would love to have this feature

**Comment by maheshmnj** [comment]
I have released a new version [v0.8.0](https://pub.dev/packages/searchfield/changelog) with a new property `onSearchTextChanged`. It must return a `<SearchFieldListItem>[]` to filter suggestions.

```dart
SearchField(
          onSearchTextChanged: (query) {
            final filter = suggestions
                .where((element) =>
                    element.toLowerCase().contains(query.toLowerCase()))
                .toList();
            return filter.map((e) => SearchFieldListItem<String>(e)).toList();
          },
          key: const Key('searchfield'),
          hint: 'Search by country name',
          suggestions:
              suggestions.map((e) => SearchFieldListItem<String>(e)).toList(),
          focusNode: focus,
          suggestionState: Suggestion.expand,
          onSuggestionTap: (SearchFieldListItem<String> x) {
            focus.unfocus();
          },
        ),
```

**Comment by arifikhsan** [comment]
cool @maheshmnj, I will check it later
